services/src/main/java/org/keycloak/protocol/ciba/decoupledauthn/DecoupledAuthenticationProviderBase.java

```diff
@@ -0,0 +1,263 @@
+import org.keycloak.protocol.ciba.utils.DecoupledAuthnResult;
+import org.keycloak.protocol.ciba.utils.DecoupledAuthnResultParser;
+import org.keycloak.protocol.ciba.utils.DecoupledAuthStatus;
+import org.keycloak.protocol.oidc.OIDCLoginProtocol;
+import org.keycloak.protocol.oidc.ext.OIDCExtProvider;
+import org.keycloak.protocol.oidc.utils.AuthorizeClientUtil;
+import org.keycloak.services.CorsErrorResponseException;
+import org.keycloak.services.Urls;
+import org.keycloak.services.managers.AuthenticationManager;
+import org.keycloak.services.resources.Cors;
+import org.keycloak.sessions.AuthenticationSessionModel;
+import org.keycloak.sessions.RootAuthenticationSessionModel;
+
+public abstract class DecoupledAuthenticationProviderBase implements DecoupledAuthenticationProvider, OIDCExtProvider {
+
+    private static final Logger logger = Logger.getLogger(DecoupledAuthenticationProviderBase.class);
+
+    protected KeycloakSession session;
+    protected EventBuilder event;
+
+    protected MultivaluedMap<String, String> formParams;
+
+    protected RealmModel realm;
+
+    protected ClientModel client;
+    protected Map<String, String> clientAuthAttributes;
+
+    @Context
+    protected HttpHeaders headers;
+    @Context
+    protected HttpRequest httpRequest;
+    @Context
+    protected HttpResponse httpResponse;
+    @Context
+    protected ClientConnection clientConnection;
+
+    protected Cors cors;
+
+    public DecoupledAuthenticationProviderBase(KeycloakSession session) {
+        this.session = session;
+        realm = session.getContext().getRealm();
+    }
+
+    @Override
+    public void setEvent(EventBuilder event) {
+        this.event = event;
+    }
+
+    @Override
+    public void close() {
+    }
+
+    @Path("/")
+    @POST
+    @Consumes(MediaType.APPLICATION_FORM_URLENCODED)
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response processDecoupledAuthnResult() {
+        event.event(EventType.LOGIN);
+        headers = session.getContext().getContextObject(HttpHeaders.class);
+        httpRequest = session.getContext().getContextObject(HttpRequest.class);
+        httpResponse = session.getContext().getContextObject(HttpResponse.class);
+        clientConnection = session.getContext().getContextObject(ClientConnection.class);
+
+        cors = Cors.add(httpRequest).auth().allowedMethods("POST").auth().exposedHeaders(Cors.ACCESS_CONTROL_ALLOW_METHODS);
+        formParams = httpRequest.getDecodedFormParameters();
+
+        checkSsl();
+        checkRealm();
+        // here Client Model of Decoupled Authentication Server is set to this class member "client".
+        // this will be overridden by Client Model of CD(Consumption Device) in verifyDecoupledAuthnResult().
+        checkClient();
+        logger.info(" client_id = " + client.getClientId());
+
+        MultivaluedMap<String, Object> outputHeaders = httpResponse.getOutputHeaders();
+        outputHeaders.putSingle("Cache-Control", "no-store");
```

Comment:
Prefer using constants from `org.jboss.resteasy.util.HttpHeaderNames`.

Also, the `Response` type from JAX-RS should give you methods to set the cache-control as well as any other header, without having to deal with the HttpResponse directly.

Reply:
@pedroigor Yes. I will fix codes as follows as you suggested.
```
        return cors.builder(Response.ok("", MediaType.APPLICATION_JSON_TYPE)
                .header(HttpHeaderNames.CACHE_CONTROL, "no-store")
                .header(HttpHeaderNames.PRAGMA, "no-cache"))
                .build();
```